File: src/main/python/AoC2022_14.py
#! /usr/bin/env python3
#
# Advent of Code 2022 Day 14
#


import itertools
import sys
from dataclasses import dataclass
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST = """\
498,4 -> 498,6 -> 496,6
503,4 -> 502,4 -> 502,9 -> 494,9
"""

Position = tuple[int, int]
SOURCE: Position = (500, 0)


@dataclass(frozen=True)
class Cave:
    rocks: set[Position]
    max_x: int
    max_y: int

    @classmethod
    def from_inputs(cls, inputs: InputData) -> Self:
        rocks = set[Position]()
        max_x = 0
        max_y = 0
        for line in inputs:
            c = [
                list(map(int, p.split(",")))
                for p in line.strip().split(" -> ")
            ]
            for (x_1, y_1), (x_2, y_2) in itertools.pairwise(c):
                x1, x2 = sorted([x_1, x_2])
                y1, y2 = sorted([y_1, y_2])
                for x in range(x1, x2 + 1):
                    for y in range(y1, y2 + 1):
                        rocks.add((x, y))
                        max_x = max(max_x, x)
                        max_y = max(max_y, y)
        return cls(rocks, max_x, max_y)

    def get_occupied(self) -> list[list[bool]]:
        occupied = [
            [False] * (self.max_x + 150) for _ in range(self.max_y + 2)
        ]
        for x, y in self.rocks:
            occupied[y][x] = True
        return occupied


Input = Cave
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Cave.from_inputs(input_data)

    def solve(self, occupied: list[list[bool]], max_y: int) -> int:  # noqa:C901
        def drop() -> Position | None:
            curr = SOURCE
            while True:
                p = curr
                curr_x, curr_y = curr
                for dx, dy in [(0, 1), (-1, 1), (1, 1)]:
                    try:
                        if not occupied[curr_y + dy][curr_x + dx]:
                            curr = (curr_x + dx, curr_y + dy)
                            break
                    except IndexError:
                        pass
                if curr == p:
                    return curr
                if curr[1] > max_y:
                    return None

        cnt = 0
        while True:
            p = drop()
            if p is None:
                break
            occupied[p[1]][p[0]] = True
            cnt += 1
            if p == SOURCE:
                break
        return cnt

    def part_1(self, cave: Input) -> Output1:
        return self.solve(cave.get_occupied(), cave.max_y)

    def part_2(self, cave: Input) -> Output2:
        return self.solve(cave.get_occupied(), cave.max_y + 2)

    @aoc_samples(
        (
            ("part_1", TEST, 24),
            ("part_2", TEST, 93),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2022, 14)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2022_14.java
import static com.github.pareronia.aoc.Utils.toAString;
import static java.util.stream.Collectors.toList;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2022_14
                extends SolutionBase<AoC2022_14.Cave, Integer, Integer> {
    
    public static final Position SOURCE = Position.of(500, 0);
    private static final char START = '+';
    private static final char EMPTY = ' ';
    private static final char ROCK = '▒';
    private static final char SAND = 'o';
    
    private final List<Listener> listeners = new ArrayList<>();

    private AoC2022_14(final boolean debug) {
        super(debug);
    }
    
    public static final AoC2022_14 create() {
        return new AoC2022_14(false);
    }

    public static final AoC2022_14 createDebug() {
        return new AoC2022_14(true);
    }
    
    public void addListener(final Listener listener) {
        this.listeners.add(listener);
    }
    
    Optional<Position> drop(final boolean[][] occupied, final int maxY) {
        Position curr = SOURCE;
        while (true) {
            final int y = curr.getY() + 1;
            final Position p = curr;
            for (final int x : List.of(0, -1 , 1)) {
                try {
                    if (!occupied[y][curr.getX() + x]) {
                        curr = Position.of(curr.getX() + x, y);
                        break;
                    }
                } catch (final ArrayIndexOutOfBoundsException e) {
                }
            }
            if (curr.equals(p)) {
                return Optional.of(curr);
            }
            if (curr.getY() > maxY) {
                return Optional.empty();
            }
        }
    }
    
    private int solve(final Cave cave, final int maxY) {
        this.listeners.forEach(l -> l.start(cave));
        int cnt = 0;
        while (true) {
            final Optional<Position> p = drop(cave.occupied, maxY);
            if (p.isPresent()) {
                cave.occupied[p.get().getY()][p.get().getX()] = true;
                this.listeners.forEach(l -> l.stateUpdated(cave));
                cnt++;
                if (p.get().equals(SOURCE)) {
                    break;
                }
            } else {
                break;
            }
        }
        this.listeners.forEach(Listener::close);
        return cnt;
    }
    
    void draw(final Cave cave) {
        if (!this.debug) {
            return;
        }
        final var statsX = IntStream.range(0, cave.occupied.length)
            .flatMap(y -> IntStream.range(0, cave.occupied[y].length)
            .filter(x -> cave.occupied[y][x]))
            .summaryStatistics();
        IntStream.rangeClosed(SOURCE.getY(), cave.maxY).forEach(y -> {
            final var line = IntStream.rangeClosed(statsX.getMin(), statsX.getMax())
                    .mapToObj(x -> {
                        final Position pos = Position.of(x, y);
                        if (cave.rocks.contains(pos)) {
                            return ROCK;
                        } else if (cave.occupied[pos.getY()][pos.getX()]) {
                            return SAND;
                        } else if (SOURCE.equals(pos)) {
                            return START;
                        }
                        return EMPTY;
                    })
                    .collect(toAString());
            log(line);
        });
    }
    
    @Override
    protected Cave parseInput(final List<String> inputs) {
        return Cave.fromInput(inputs);
    }

    @Override
    public Integer solvePart1(final Cave cave) {
        final int ans = solve(cave, cave.maxY);
        draw(cave);
        return ans;
    }

    @Override
    public Integer solvePart2(final Cave cave) {
        final Cave newCave = Cave.withRocks(cave.rocks);
        final int ans = solve(newCave, cave.maxY + 2);
        draw(newCave);
        return ans;
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "24"),
        @Sample(method = "part2", input = TEST, expected = "93"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2022_14.create().run();
    }

    protected static final String TEST = """
        498,4 -> 498,6 -> 496,6
        503,4 -> 502,4 -> 502,9 -> 494,9
        """;

    record Cave(Set<Position> rocks, boolean[][] occupied, int maxY) {
        
        public static Cave withRocks(final Set<Position> rocks) {
            final int maxX = rocks.stream().mapToInt(Position::getX).max().getAsInt();
            final int maxY = rocks.stream().mapToInt(Position::getY).max().getAsInt();
            final var occupied = new boolean[maxY + 2][maxX + 150];
            rocks.stream().forEach(p -> occupied[p.getY()][p.getX()] = true);
            return new Cave(rocks, occupied, maxY);
        }

        public static Cave fromInput(final List<String> input) {
            final Set<Position> rocks = new HashSet<>();
            for (final String line : input) {
                final var splits = line.split(" -> ");
                IntStream.range(1, splits.length).forEach(i -> {
                    final var splits1 = splits[i - 1].split(",");
                    final var splits2 = splits[i].split(",");
                    final var xs = List.of(splits1, splits2).stream()
                            .map(a -> a[0])
                            .map(Integer::parseInt)
                            .sorted()
                            .collect(toList());
                    final var ys = List.of(splits1, splits2).stream()
                            .map(a -> a[1])
                            .map(Integer::parseInt)
                            .sorted()
                            .collect(toList());
                    IntStream.rangeClosed(xs.get(0), xs.get(1)).forEach(x ->
                        IntStream.rangeClosed(ys.get(0), ys.get(1)).forEach(y ->
                            rocks.add(Position.of(x,  y))));
                });
            }
            return Cave.withRocks(rocks);
        }
    }
    
    public interface Listener {
        void start(Cave cave);
        
        void stateUpdated(Cave cave);
        
        default void close() {}
    }
}
